Download shader as a ZIP including a readme.txt

cxw42 opened the pull request:
This adds a one-click ZIP download.  The ZIP includes the JSON file, and also a readme.txt with the title, description, and tags.

Reason: to more easily upload ZIPs to scene.org.  I can hit "Export ZIP" and have a file ready to upload, unless the readme needs hand-tweaking.

I also added a license file, since I pulled in JSZip and node-sanitize-filename.  See also #46.

Thanks for considering this request!

cxw42 commented:
@patuwwy as we discussed, I will add to the popup a checkbox to enable ZIP support, including this feature, then ping you when it's ready.  Thanks!

patuwwy commented:
#47

patuwwy commented on manifest.json:
```diff
@@ -36,6 +36,8 @@
         "shadertoy-plugin-profile.js",
-        "shadertoy-plugin-home.js"
+        "shadertoy-plugin-home.js",
+        "/js/jszip-3.1.5.js",
```
please move additional files to "lib" folder

/lib/jszip-3.1.5.js
/lib/node-sanitize-filename.js

there are js files in root folder so adding "js" folder for only 2 new files is not good idea IMO

patuwwy commented on shadertoy-plugin.js:
```diff
@@ -90,6 +96,14 @@
+     */
+    ToyPlugCommon.prototype.downloadJson = function downloadJson(filename, data) {
+        var blob = new Blob([data], {type: 'application/json'});
+        this.downloadBlob(filename, blob);
```
add empty line after var

patuwwy commented on shadertoy-plugin.js:
```diff
@@ -381,6 +400,121 @@
+
+            // Get names we'll need
+            var id = gShaderToy.mInfo.id;
+            if (id == '-1') {
```
id === '-1'

patuwwy commented on shadertoy-plugin.js:
```diff
@@ -381,6 +400,121 @@
+            }
+
+            var friendly_name =
+                gShaderToy.mInfo.name ||
```
wrong line break + wrong indentation. this could be in single line

patuwwy commented on shadertoy-plugin.js:
```diff
@@ -381,6 +400,121 @@
+            // Get the tags from the tag box, with a fallback to the
+            // ShaderToy info.
+            var tags = [];
+            if(shaderToyElements.shaderTags &&
```
please insert space after "if"

patuwwy commented on shadertoy-plugin.js:
```diff
@@ -381,6 +400,121 @@
+            if(tags.length == 0) tags = gShaderToy.mInfo.tags;
+
+            // Add the tags to the README
+            if(Array.isArray(tags) && tags.length != 0) {
```
insert space after if

patuwwy commented on shadertoy-plugin.js:
```diff
@@ -381,6 +400,121 @@
+            // to the ShaderToy info
+            var description;
+            if(shaderToyElements.shaderDescription &&
+                            shaderToyElements.shaderDescription.value) {
```
wrong indentation

patuwwy commented on shadertoy-plugin.js:
```diff
@@ -381,6 +400,121 @@
+            }
+
+            // Add the description to the README
+            if(description) {
```
missing space after if

patuwwy commented on shadertoy-plugin.js:
```diff
@@ -381,6 +400,121 @@
+            }
+
+            // Add a usage reminder
+            readme += '\n' + '-'.repeat(78) + '\n';
```
please use template for readme contents,

cxw42 replied:
Which type of template?

 - ES6 template literals (Chrome 41+, according to MDN)
 - a string literal (or XHR from a separate file in the extension) I can regex-replace, 
 - a template engine (I doubt it)
 - something else

patuwwy replied:
i meant smth like this:

```
README_CONTENT_TEMPLATE = '---------------- {shaderName} --------\n' +
    '----{shaderTags} ------\n {shaderDescription}';
```

then when building readme contents:

```
var readmeContent = '';
...

readmeContent = README_CONTENT_TEMPLATE
    .replace('{shaderName}', getShaderName())
    .replace('{shaderTags}', getShaderTags())
    .replace.....
```

please look [here](https://github.com/patuwwy/ShaderToy-Chrome-Plugin/blob/455dd154ba1855b1e3436c00e4a5110d3c877882/shadertoy-plugin-profile.js#L32)

cxw42 replied:
Thanks for the link - I hadn't seen that.  Since we will have common code, OK if I abstract out a template-applying function?  Like `render_template(README_TEMPLATE,{shaderName: getShaderName(), shaderTags: getShaderTags()})` or some such?

patuwwy commented on contentscript.js:
```diff
@@ -21,32 +21,40 @@
+        /**
+         * ZIP export script filename, used by the main extension script.
+         *
+         * @type {string}
```
`* @const {string}`

cxw42 commented:
@patuwwy Thanks for the comments - will do!  I will send you a message when it's complete - right now it loads the scripts twice.

patuwwy commented on contentscript.js:
```diff
@@ -89,10 +97,46 @@
+
+            // Enable button on the profile page
+            elem = document.getElementById('dl-all-shader-zip');
+            if(elem) elem.style.display = 'block';
```
IFIK this PR is about downloading ZIP with single shader :)

cxw42 replied:
That is a good point.  However, I don't want to implement this in a way that won't also work for the profile page.  I will test both, then remove the profile-page material when I send you the complete version for your review.  That will make it easy to finish #49 after this is merged.

patuwwy commented on contentscript.js:
```diff
@@ -153,6 +203,32 @@
+     * when the Enable ZIP checkbox is selected.
+     */
+    function loadZipScripts() {
+        if(document.getElementById('script-jszip')) return;
```
space after if, brackets missing

patuwwy commented on contentscript.js:
```diff
@@ -153,6 +203,32 @@
+            // Don't reload if the scripts are already there
+
+        chrome.storage.sync.get('enableZip', function onSync(items) {
+            if(items && items.enableZip) {
```
missing space

patuwwy commented on contentscript.js:
```diff
@@ -166,8 +242,9 @@
-            loadScript(HOME_EXTENSION_FILENAME);
+        if (document.location.href.match(/shadertoy.com\/?$/)) {
+            //loadScript(HOME_EXTENSION_FILENAME);
+            // There's nothing there yet!
```
please change only PR related code

cxw42 commented:
I think I have addressed all the comments.  Next time I am able to work on this I will cherry-pick and squash onto the new master, and then force-push the resulting single commit into this PR.

cxw42 commented:
@patuwwy Here you go!  3805429 is a single commit on 033f27e, which is `master` as I write this.  I believe it is ready for your review.  I did not update the changelog or the version number in manifest.json; let me know if you would like me to do either.

patuwwy commented:
Published
